pytorch_lightning/callbacks/finetuning.py

```diff
@@ -0,0 +1,221 @@
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+BN_TYPES = (torch.nn.BatchNorm1d, torch.nn.BatchNorm2d, torch.nn.BatchNorm3d)
+
+
+def _make_trainable(module: Module) -> None:
+    """Unfreezes a given module.
+    Args:
+        module: The module to unfreeze
+    """
+    for param in module.parameters():
+        param.requires_grad = True
+    module.train()
+
+
+def _recursive_freeze(module: Module,
+                      train_bn: bool = True) -> None:
+    """Freezes the layers of a given module.
+    Args:
+        module: The module to freeze
+        train_bn: If True, leave the BatchNorm layers in training mode
+    """
+    children = list(module.children())
+    if not children:
+        if not (isinstance(module, BN_TYPES) and train_bn):
+            for param in module.parameters():
+                param.requires_grad = False
+            module.eval()
+        else:
+            # Make the BN layers trainable
+            _make_trainable(module)
+    else:
+        for child in children:
+            _recursive_freeze(module=child, train_bn=train_bn)
+
+
+def filter_params(module: Module,
+                  train_bn: bool = True) -> Generator:
+    """Yields the trainable parameters of a given module.
+    Args:
+        module: A given module
+        train_bn: If True, leave the BatchNorm layers in training mode
+    Returns:
+        Generator
+    """
+    children = list(module.children())
+    if not children:
+        if not (isinstance(module, BN_TYPES) and train_bn):
+            for param in module.parameters():
+                if param.requires_grad:
+                    yield param
+    else:
+        for child in children:
+            for param in filter_params(module=child, train_bn=train_bn):
+                yield param
+
+
+def multiplicative(epoch):
+    return 2
+
+
+class BaseFinetuningCallback(Callback):
+
+    r"""
+    BaseFinetuningCallback.
+    Overrides the finetunning_function and add your own logic there.
+
+    """
+    @staticmethod
+    def freeze(module: Module, train_bn: bool = True) -> None:
+        """Freezes the layers up to index n (if n is not None).
+        Args:
+            module: The module to freeze (at least partially)
+            train_bn: If True, leave the BatchNorm layers in training mode
+        """
+        for mod in module.parameters():
+            if (isinstance(mod, BN_TYPES) and train_bn):
+                _make_trainable(mod)
+            else:
+                mod.requires_grad = False
+
+    @staticmethod
+    def unfreeze_and_add_param_group(
+        module: Module,
+        optimizer: Optimizer,
+        lr: Optional[float] = None,
+        train_bn: bool = True
+    ):
+        """Unfreezes a module and adds its parameters to an optimizer."""
+        _make_trainable(module)
+        params_lr = optimizer.param_groups[0]['lr'] if lr is None else float(lr)
+        denom_lr = 10. if lr is None else 1.
```

Comment:
why 10?

Reply:
It was a default. Not special reason. Let me make it an argument.

Reply:
Done !